core/utils/async_utils.cpp

```diff
@@ -221,205 +227,336 @@ void read_write_mutex::unlock(bool exclusive_only /*= false*/) {
       max_threads_ = std::numeric_limits<size_t>::min();
-  }
-  else {
+    } else {
       max_threads_ = max_threads;
-  }
+    }
 
-  // create extra thread if all threads are busy and can grow pool
-  if (State::ABORT != state_ && !queue_.empty() && active_ == pool_.size() && pool_.size() < max_threads_) {
-    pool_.emplace_back([](thread_pool* pool)->void{ pool->run(); }, this);
+    if (State::ABORT != state_) {
+      maybe_spawn_worker();
+    }
   }
 
-  cond_.notify_all(); // wake any idle threads if they need termination
+  state.cond.notify_all(); // wake any idle threads if they need termination
 }
 
-bool thread_pool::run(std::function<void()>&& fn) {
-  std::lock_guard<decltype(lock_)> lock(lock_);
-
-  if (State::RUN != state_) {
-    return false; // pool not active
+bool thread_pool::run(std::function<void()>&& fn, clock_t::duration delay /*=0*/) {
+  if (!fn) {
+    return false;
   }
 
-  queue_.emplace(std::move(fn));
-  cond_.notify_one();
+  auto& state = *shared_state_;
 
-  // create extra thread if all threads are busy and can grow pool
-  if (active_ == pool_.size() && pool_.size() < max_threads_) {
-    pool_.emplace_back([](thread_pool* pool)->void{ pool->run(); }, this);
+  {
+    auto lock = make_lock_guard(state.lock);
+
+    if (State::RUN != state_) {
+      return false; // pool not active
+    }
+
+    queue_.emplace(std::move(fn), clock_t::now() + delay);
+
+    try {
+      maybe_spawn_worker();
+    } catch (...) {
+      if (pool_.empty()) {
+        // failed to spawn a thread to execute a task
+        queue_.pop();
+        return false;
+      }
+    }
   }
 
+  state.cond.notify_one();
+
   return true;
 }
 
 void thread_pool::stop(bool skip_pending /*= false*/) {
-  std::unique_lock<decltype(lock_)> lock(lock_);
+  state_ = skip_pending ? State::ABORT : State::FINISH;
+
+  decltype(queue_) empty;
+  {
+    auto& state = *shared_state_;
+    auto lock = make_unique_lock(state.lock);
 
-  if (State::RUN == state_) {
-    state_ = skip_pending ? State::ABORT : State::FINISH;
+    // wait for all threads to terminate
+    while (!pool_.empty()) {
+      state.cond.notify_all(); // wake all threads
+      state.cond.wait_for(lock, 100ms);
+    }
+
+    queue_.swap(empty);
   }
+}
 
-  // wait for all threads to terminate
-  while(!pool_.empty()) {
-    cond_.notify_all(); // wake all threads
-    cond_.wait_for(lock, std::chrono::milliseconds(100));
+void thread_pool::limits(size_t max_threads, size_t max_idle) {
+  auto& state = *shared_state_;
+
+  {
+    auto lock = make_lock_guard(state.lock);
+
+    max_threads_ = max_threads;
+    max_idle_ = max_idle;
+
+    if (State::ABORT != state_) {
+      maybe_spawn_worker();
+    }
+  }
+
+  state.cond.notify_all(); // wake any idle threads if they need termination
+}
+
+bool thread_pool::maybe_spawn_worker() {
+  assert(!shared_state_->lock.try_lock()); // lock must be held
+  assert(state_ != State::ABORT);
```

Comment:
This assert still could fire in case of following events:
Thread1: limits(X,Y)->check in line 396 (passed, state is RUN)--------------------> enter maybe_spawn_worker--assertion!
Thread2: stop(true)----------------------------------------------->line 370 state set to ABORT-------------------------------

Reply:
But IMO having ABORT here just creates one additional worker. No harm in general.